tornado/testing.py

```diff
@@ -499,13 +506,16 @@ def wrap(f):
         @functools.wraps(f)
         def pre_coroutine(self, *args, **kwargs):
             result = f(self, *args, **kwargs)
-            if isinstance(result, GeneratorType):
+            if isinstance(result, GeneratorType) or iscoroutine(result):
                 self._test_generator = result
             else:
                 self._test_generator = None
             return result
 
-        coro = gen.coroutine(pre_coroutine)
+        if iscoroutinefunction(f):
```

Comment:
On second thought, wrapping a native coroutine in `gen.coroutine` doesn't work, is this a bug in `gen.coroutine`?:
```python
from tornado import gen, ioloop

io_loop = ioloop.IOLoop.current()

async def coro_fn():
    global finished
    finished = True

finished = False
# Works, prints "None".
print(io_loop.run_sync(coro_fn))
assert finished

finished = False
# Fails, prints "<coroutine object coro_fn>".
print(io_loop.run_sync(gen.coroutine(coro_fn)))
assert finished
```
In the wrapping process, `_make_coroutine_wrapper` calls `types.coroutine(func)`, which as expected returns the native coroutine unchanged. Then `_make_coroutine_wrapper` returns the inner `wrapper` function.

When `wrapper` executes, it does:
```python
result = func(*args, **kwargs)
```
...and `result` is the coroutine object. So far the coroutine has not actually been executed (since it hasn't been awaited). `wrapper` checks if `isinstance(result, GeneratorType)`, which is **false**, so it sets the coroutine object `result` as its return value and quits.

We either need to keep this `iscoroutinefunction` check here in testing.py or (more likely?) `gen.coroutine` isn't working as intended, and it needs another branch after `isinstance(result, GeneratorType)` to do something special with native coroutines.